File: halmd/mdsim/host/forces/lennard_jones.cpp
/*
 * Copyright © 2008-2010  Peter Colberg and Felix Höfling
 *
 * This file is part of HALMD.
 *
 * HALMD is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <boost/numeric/ublas/io.hpp>
#include <cmath>
#include <string>

#include <halmd/io/logger.hpp>
#include <halmd/io/utility/hdf5.hpp>
#include <halmd/mdsim/host/forces/lennard_jones.hpp>
#include <halmd/utility/lua_wrapper/lua_wrapper.hpp>

using namespace boost;
using namespace boost::assign;
using namespace boost::numeric::ublas;
using namespace std;

namespace halmd
{
namespace mdsim { namespace host { namespace forces
{

/**
 * Assemble module options
 */
template <typename float_type>
void lennard_jones<float_type>::options(po::options_description& desc)
{
    desc.add_options()
        ("cutoff", po::value<boost::array<float, 3> >()->default_value(default_cutoff()),
         "truncate potential at cutoff radius")
        ("epsilon", po::value<boost::array<float, 3> >()->default_value(default_epsilon()),
         "potential well depths AA,AB,BB")
        ("sigma", po::value<boost::array<float, 3> >()->default_value(default_sigma()),
         "collision diameters AA,AB,BB")
        ;
}

/**
 * Register option value types with Lua
 */
static __attribute__((constructor)) void register_option_converters()
{
    register_any_converter<boost::array<float, 3> >();
}

/**
 * Write module parameters to HDF5 group
 */
template <typename float_type>
void lennard_jones<float_type>::write_parameters(H5::Group const& group) const
{
    h5xx::write_attribute(group, "epsilon", epsilon_.data());
    h5xx::write_attribute(group, "sigma", sigma_.data());
    h5xx::write_attribute(group, "cutoff", r_cut_sigma_.data());
}

/**
 * Initialise Lennard-Jones potential parameters
 */
template <typename float_type>
lennard_jones<float_type>::lennard_jones(
    unsigned ntype
  , array<float, 3> const& cutoff
  , array<float, 3> const& epsilon
  , array<float, 3> const& sigma
)
  // allocate potential parameters
  : epsilon_(scalar_matrix<float_type>(ntype, ntype, 1))
  , sigma_(scalar_matrix<float_type>(ntype, ntype, 1))
  , r_cut_sigma_(ntype, ntype)
  , r_cut_(ntype, ntype)
  , rr_cut_(ntype, ntype)
  , sigma2_(ntype, ntype)
  , en_cut_(scalar_matrix<float_type>(ntype, ntype, 0))
{
    // FIXME support any number of types
    for (unsigned i = 0; i < std::min(ntype, 2U); ++i) {
        for (unsigned j = i; j < std::min(ntype, 2U); ++j) {
            epsilon_(i, j) = epsilon[i + j];
            sigma_(i, j) = sigma[i + j];
            r_cut_sigma_(i, j) = cutoff[i + j];
        }
    }

    // precalculate derived parameters
    for (unsigned i = 0; i < ntype; ++i) {
        for (unsigned j = i; j < ntype; ++j) {
            r_cut_(i, j) = r_cut_sigma_(i, j) * sigma_(i, j);
            rr_cut_(i, j) = std::pow(r_cut_(i, j), 2);
            sigma2_(i, j) = std::pow(sigma_(i, j), 2);
            // energy shift due to truncation at cutoff length
            en_cut_(i, j) = (*this)(rr_cut_(i, j), i, j).second;
        }
    }

    LOG("potential well depths: ε = " << epsilon_);
    LOG("potential core width: σ = " << sigma_);
    LOG("potential cutoff length: r_c = " << r_cut_sigma_);
    LOG("potential cutoff energy: U = " << en_cut_);
}

template <typename float_type>
void lennard_jones<float_type>::luaopen(lua_State* L)
{
    using namespace luabind;
    module(L, "halmd_wrapper")
    [
        namespace_("mdsim")
        [
            namespace_("host")
            [
                namespace_("forces")
                [
                    class_<lennard_jones, shared_ptr<lennard_jones> >(module_name())
                        .def(constructor<
                            unsigned
                          , array<float, 3> const&
                          , array<float, 3> const&
                          , array<float, 3> const&
                        >())
                        .def("write_parameters", &lennard_jones::write_parameters)
                        .scope
                        [
                            def("options", &lennard_jones::options)
                        ]
                ]
            ]
        ]
    ];
}

static __attribute__((constructor)) void register_lua()
{
#ifndef USE_HOST_SINGLE_PRECISION
    typedef double float_type;
#else
    typedef float float_type;
#endif

    lua_wrapper::register_(0) //< distance of derived to base class
    [
        &lennard_jones<float_type>::luaopen
    ];

    lua_wrapper::register_(2) //< distance of derived to base class
    [
        &pair_trunc<3, float_type, lennard_jones<float_type> >::luaopen
    ]
    [
        &pair_trunc<2, float_type, lennard_jones<float_type> >::luaopen
    ];
}

// explicit instantiation
#ifndef USE_HOST_SINGLE_PRECISION
template class lennard_jones<double>;
template class pair_trunc<3, double, lennard_jones<double> >;
template class pair_trunc<2, double, lennard_jones<double> >;
#else
template class lennard_jones<float>;
template class pair_trunc<3, float, lennard_jones<float> >;
template class pair_trunc<2, float, lennard_jones<float> >;
#endif

}}} // namespace mdsim::host::forces

} // namespace halmd

File: lua/halmd/mdsim/forces/lennard_jones.lua
--
-- Copyright © 2010  Peter Colberg and Felix Höfling
--
-- This file is part of HALMD.
--
-- HALMD is free software: you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation, either version 3 of the License, or
-- (at your option) any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with this program.  If not, see <http://www.gnu.org/licenses/>.
--

require("halmd.modules")

-- grab environment
local lennard_jones_wrapper = {
    host = halmd_wrapper.mdsim.host.forces.lennard_jones
}
if halmd_wrapper.mdsim.gpu then
    lennard_jones_wrapper.gpu = halmd_wrapper.mdsim.gpu.forces.lennard_jones
end

local mdsim = {
    core = require("halmd.mdsim.core")
}
local device = require("halmd.device")
local po = halmd_wrapper.po
local assert = assert

module("halmd.mdsim.forces.lennard_jones", halmd.modules.register)

--
-- construct Lennard-Jones module
--
function new(args)
    local cutoff = args.cutoff or { 2.5, 2.5, 2.5 } -- default value
    local epsilon = args.epsilon or { 1.0, 1.5, 0.5 } -- default value
    local sigma = args.sigma or { 1.0, 0.8, 0.88 } -- default value

    local core = mdsim.core()
    local particle = assert(core.particle)

    local lennard_jones
    if device() then
        lennard_jones = assert(lennard_jones_wrapper.gpu)
    else
        lennard_jones = assert(lennard_jones_wrapper.host)
    end
    return lennard_jones(particle.ntype, cutoff, epsilon, sigma)
end

--
-- assemble module options
--
-- @param desc po.options_description
--
function options(desc)
    desc:add("cutoff", po.float_array(), "truncate potential at cutoff radius")
    desc:add("epsilon", po.float_array(), "potential well depths")
    desc:add("sigma", po.float_array(), "collision diameters")
    -- FIXME desc:add("smooth", po.float_array(), "C²-potential smoothing factor")
end
